include/ygm/container/detail/bag_impl.hpp

```diff
@@ -47,11 +89,74 @@ class bag_impl {
     return m_comm.all_reduce_sum(m_local_bag.size());
   }
 
+  size_t local_size() {
+    return m_local_bag.size();
+  }
+
+
+  void rebalance() {
+    m_comm.barrier();
+
+    // Find current rank's prefix val and desired target size
+    size_t prefix_val = ygm::prefix_sum(local_size(), m_comm);
+    size_t target_size = std::ceil((size() * 1.0) / m_comm.size());
+
+    // Init to_send array where index is dest and value is the num to send
+    //int to_send[m_comm.size()] = {0};
+    std::unordered_map<size_t, size_t> to_send;
+    
+    for (size_t i = 0; i < local_size(); i++) {
+      size_t idx = prefix_val + i;
+      size_t target_rank = idx / target_size;
+      if (target_rank != m_comm.rank()) 
+        to_send[target_rank]++;
+    }
+    m_comm.barrier();
+
+    // Build and send bag indexes as calculated by to_send
+    for (size_t r = 0; r < m_comm.size(); r++)
```

Comment:
Iterating over the `std::unordered_map` in this way will add elements until its size is O(p), which gets back to the same issue as using a `std::vector`. Instead, iterate over only the entries that exist, i.e. `for (const auto &kv : to_send)`. Also, please explicitly add `{ }` around code blocks. It makes the code easier to parse and gets rid of an avenue for future bugs.